### PR TITLE
Implement determistic resolver selection and initial telemetry ingestion

### PR DESCRIPTION
**Pull request by sudheesh001**
Signed-off-by: Sudheesh Singanamalla <sudheesh@cloudflare.com>

### CODE REVIEW

File: target.go [hunk 1]
```diff
@@ -186,25 +225,50 @@ func (s *targetServer) obliviousQueryHandler(w http.ResponseWriter, r *http.Requ
...
 
-	packedResponse, err := s.resolveQuery(query)
+	queryParseAndDecryptionCompleteTime := time.Now().UnixNano()
+	timestamp.TargetQueryDecryptionTime = queryParseAndDecryptionCompleteTime
+
+	resolverChosen := s.resolver[chosenResolver]
+	packedResponse, err := s.resolveQueryUsingNS(query, resolverChosen)
 	if err != nil {
 		log.Println("Failed resolving DNS query:", err)
 		http.Error(w, http.StatusText(http.StatusInternalServerError), http.StatusInternalServerError)
```
Comment (chris-wood) [inline target.go:235]: Should we also log an error here?

File: target.go [hunk 2]
```diff
@@ -88,7 +89,37 @@ func (s *targetServer) resolveQuery(query *dns.Msg) ([]byte, error) {
 	}
 
 	start := time.Now()
-	response, err := s.resolver.resolve(query)
+	response, err := s.resolver[0].resolve(query)
```
Comment (chris-wood) [inline target.go:92]: ~~This doesn't choose randomly from the list of resolvers, does it?~~

Ah, this seems to be an older function. Maybe we should remove it?

File: target.go [hunk 3]
```diff
@@ -35,8 +35,9 @@ import (
 
 type targetServer struct {
 	verbose     bool
-	resolver    *targetResolver
+	resolver    []*targetResolver
 	odohKeyPair odoh.ObliviousDNSKeyPair
+	telemetryClient *telemetry
 }
 
 func decodeDNSQuestion(encodedMessage []byte) (*dns.Msg, error) {
```
Comment (chris-wood) [inline target.go:43]: Are we still using this function? If not, can we delete it?

File: target.go [hunk 4]
```diff
@@ -88,7 +89,37 @@ func (s *targetServer) resolveQuery(query *dns.Msg) ([]byte, error) {
...
+	}
+
+	if s.verbose {
+		log.Printf("Answer=%s elapsed=%s\n", packedResponse, elapsed.String())
+	}
+
+	return packedResponse, err
+}
+
+func (s *targetServer) resolveQueryUsingNS(query *dns.Msg, resolver *targetResolver) ([]byte, error) {
```
Comment (chris-wood) [inline target.go:108]:
```suggestion
func (s *targetServer) resolveQueryWithResolver(query *dns.Msg, resolver *targetResolver) ([]byte, error) {
```

Comment (chris-wood) [reply target.go:108]: (It's not really a name server.)

File: target.go [hunk 5]
```diff
@@ -173,11 +204,19 @@ func (s *targetServer) createObliviousResponseForQuery(query *odoh.ObliviousDNSQ
 }
 
 func (s *targetServer) obliviousQueryHandler(w http.ResponseWriter, r *http.Request) {
+	requestReceivedTime := time.Now()
```
Comment (chris-wood) [inline target.go:207]: Can we put all metrics collection code behind a flag? Or at least put the elastic search ingestion behind a flag?

File: target.go [hunk 6]
```diff
@@ -186,25 +225,50 @@ func (s *targetServer) obliviousQueryHandler(w http.ResponseWriter, r *http.Requ
...
+
 	obliviousResponse, err := s.createObliviousResponseForQuery(obliviousQuery, packedResponse)
 	if err != nil {
 		log.Println("Failed creating DNS oblivious DNS response:", err)
+		timestamp.TargetAnswerEncryptionTime = 0
+		timestamp.EndTime = 0
+		exp.Timestamp = timestamp
+		exp.Status = false
+		exp.Resolver = ""
+		go s.telemetryClient.streamDataToElastic([]string{exp.serialize()})
```
Comment (chris-wood) [inline target.go:250]: Can we abstract over Elasticsearch? For example, we could simply name this function `submit`, and then the specific telemetry client could determine how and where to send the data.

Comment (chris-wood) [reply target.go:250]: Also, calling `streamDataToElastic` once every query could be costly. Should we batch them? For example, we could just do:

```suggestion
s.telemetryClient.submit(exp.serialize())
```

And internally, for the Elasticsearch telemetry client, that would queue up strings to log, and send them when done.

File: target.go (hunk 3 above)
Comment (sudheesh001) [reply target.go:44]: This is still used for DOH queries.

File: target.go (hunk 6 above)
Comment (sudheesh001) [reply target.go]: I'll do this in parallel with https://github.com/chris-wood/odoh-client/issues/17 when its being tackled on the client.

File: target.go (hunk 1 above)
Comment (sudheesh001) [reply target.go:236]: We do have a log statement before the error is returned which show up on the GAE logs

**Pull request merged**